cirq/circuits/circuit_dag.py

```diff
@@ -209,3 +209,24 @@ def findall_nodes_until_blocked(
                 remaining_dag.remove_node(node)
                 continue
             yield node
+
+    def findall_nodes_from_set(
+        self, input_set: Iterable[Unique[ops.Operation]]
+    ) -> Iterator[ops.Operation]:
+        """Finds all nodes from the given set.
+
+        Args:
+            input: The input set of Operation nodes.
+        """
+        for node in self.ordered_nodes():
+            if node in input_set:
+                yield node.val
+
+    def factorize(self) -> Iterator['cirq.CircuitDag']:
+        """Tries to factorize the underlying graph, using connected components algorithm.
+        If no factorization is possible, returns a sequence with a single element (itself).
+        """
+        for c in networkx.weakly_connected_components(self):
+            yield CircuitDag.from_ops(
```

Comment:
I think this will destroy the moment structure of the input circuit, right? Maybe this isn't necessary if we're using this functionality inside a simulator but it could be important if e.g. trying to factorize a circuit prior to running on a real device when we want to have more precise control over the timing of operations

Reply:
+1 to this. It'd be nice if we could factorize regular circuits (not just `CircuitDag`) in a way that preserves moments, so those circuits could be zipped together to recover the original circuit:
```python
cirq.Circuit.zip(*circuit.factorize()) == circuit
```

Reply:
Fair point. That was my question a while ago (considering one of the examples [here](https://github.com/quantumlib/Cirq/issues/3409#issuecomment-732647238). One way to do it could be by keeping track of the moments corresponding to operations. I'll try to get something working in that direction.

Reply:
I have addressed this, but ended up making a spin-off from the existing CircuitDag (because the existing class is already used in a few places). Could you have another look at the PR and let me know your thoughts?